# Практична робота 2. Аналіз даних. Робота даними csv
## Amazon Top 50 Bestselling Books 2009 – 2019

Набір даних із 50 найбільш продаваних книг Amazon з 2009 по 2019 рік. Містить 550 книг, дані були розділені на художню літературу та нон-фікшн за допомогою Goodreads. 

Для кожної з книг доступні 7 колонок.  
- Name - назва книги 
- Author - автор 
- User Rating - рейтинг (за 5-бальною шкалою) 
- Reviews - кількість відгуків 
- Price - ціна (у доларах станом на 2020 рік) 
- Year - рік, коли книга потрапила до рейтингу Топ-50 
- Genre - жанр 

In [1]:
import pandas as pd

1. Прочитайте csv файл (метод read_csv) 

In [2]:
df = pd.read_csv("bestsellers with categories.csv")

2. Виведіть перші 10 рядків (метод head) 

In [3]:
df.head(10)

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   
5      A Dance with Dragons (A Song of Ice and Fire)   
6  A Game of Thrones / A Clash of Kings / A Storm...   
7                     A Gentleman in Moscow: A Novel   
8      A Higher Loyalty: Truth, Lies, and Leadership   
9                          A Man Called Ove: A Novel   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction  
5       George R. R. Martin          4.4    12643     11  2011      Fiction  
6       George R. R. Martin          4.7    19735     30  2014      Fiction  
7               Amor Towles          4.7    19699     15  2017      Fiction  
8               James Comey          4.7     5983      3  2018  Non Fiction  
9           Fredrik Backman          4.6    23848      8  2016      Fiction

3. Виведіть розміри датасету (метод shape)

In [4]:
df.shape

(550, 7)

Питання: Про скільки книг зберігає дані датасет? 

Я могла би сказати сходу, що у нас 550 записів, значить дані про 550 книг. 
Але вирішила перевірити унікальні назви.

In [5]:
df.Name.nunique()

351

Бачимо, що унікальних всього 351. 
Тому вирішила подивитися на дублікати

In [6]:
df[df['Name'].duplicated()].count()

Name           199
Author         199
User Rating    199
Reviews        199
Price          199
Year           199
Genre          199
dtype: int64

Дійсно, по всім колонкам однакова кількість дублікатів. 
До того ж, ця кількість як раз 199 (+351 = 550)

### Висновок: маємо інформацію про 351 книгу

4. Змініть назви всіх колонок: назва з маленької букви, без пробілів через підкреслення. 

In [7]:
df.columns = df.columns.str.replace(" ", "_").str.lower()
df.columns

Index(['name', 'author', 'user_rating', 'reviews', 'price', 'year', 'genre'], dtype='object')

5. Вивести кількість пропусків у кожному стовпчику. 

In [8]:
df.isna().sum()

name           0
author         0
user_rating    0
reviews        0
price          0
year           0
genre          0
dtype: int64

Питання : Чи є в якихось змінних пропуски? (Так / ні) 
- Відповідь: немає пропусків

За допомогою методу unique  перевірте. Які є унікальні жанри в стовпці genre  

In [9]:
df.genre.unique()

array(['Non Fiction', 'Fiction'], dtype=object)

Питання : Які є унікальні жанри?
- Відповідь: 'Non Fiction', 'Fiction'

7. Визначте максимальну, мінімальну, медіанну ціну за допомогою методів  max, min, mean, median 

Питання: Максимальна ціна? 

Питання: Мінімальна ціна? 

Питання: Середня ціна? 

Питання: Медіанна ціна? 

In [10]:
print("Максимальна ціна:", df['price'].max())
print("Мінімальна ціна:", df['price'].min())
print("Середня ціна:", df['price'].mean())
print("Медіанна ціна:", df['price'].median())

Максимальна ціна: 105
Мінімальна ціна: 0
Середня ціна: 13.1
Медіанна ціна: 11.0


8. Виконайте наступні завдання з пошуку і сортування даних 

Питання: Який рейтинг у датасеті найвищий?  

In [11]:
max_rating = df['user_rating'].max()
print("Найвищий рейтинг у датасеті:", max_rating)

Найвищий рейтинг у датасеті: 4.9


Питання: Скільки книг мають такий рейтинг?  

In [12]:
count_max_rating_books = df[df['user_rating'] == max_rating]['name'].count()
print("Кількість книг з найвищим рейтингом:", count_max_rating_books)

Кількість книг з найвищим рейтингом: 52


Питання: Яка книга має найбільше відгуків?  

In [13]:
max_reviews_book = df.loc[df['reviews'].idxmax(), 'name']
print("Книга з найбільшою кількістю відгуків:", max_reviews_book)

Книга з найбільшою кількістю відгуків: Where the Crawdads Sing


Питання: З тих книг, що потрапили до Топ-50 у 2010 році, яка книга найдорожча ? 

In [14]:
top_50_2010 = df[(df['year'] == 2010)].nlargest(50, 'user_rating')

most_expensive_book = top_50_2010[top_50_2010['price'] == top_50_2010['price'].max()]

print("З Топ-50 у 2010 році, найдорожча книга:")
most_expensive_book

З Топ-50 у 2010 році, найдорожча книга:


name  \
272  Publication Manual of the American Psychologic...   

                                 author  user_rating  reviews  price  year  \
272  American Psychological Association          4.5     8580     46  2010   

           genre  
272  Non Fiction

Питання : Скільки книг жанру Fiction потрапили до Топ-50 у 2012 році  

In [15]:
fiction_2012 = len(df[(df['genre'] == 'Fiction') & (df['year'] == 2012)])
print("Кількість книг жанру Fiction, які потрапили до Топ-50 у 2012 році:", fiction_2012)

Кількість книг жанру Fiction, які потрапили до Топ-50 у 2012 році: 21


Питання : Скільки книг з рейтингом 4.9 потрапило до рейтингу у 2010 та 2011 роках (використовуйте | або метод isin)?  

In [16]:
top_books_2010_2011 = df[(df['user_rating'].isin([4.9])) & (df['year'].isin([2010, 2011]))]
print("Кількість книг з рейтингом 4.9, які потрапили до рейтингу у 2010 або 2011 році:", len(top_books_2010_2011))

Кількість книг з рейтингом 4.9, які потрапили до рейтингу у 2010 або 2011 році: 1


І це така книга:

In [17]:
top_books_2010_2011

name       author  \
187  Jesus Calling: Enjoying Peace in His Presence ...  Sarah Young   

     user_rating  reviews  price  year        genre  
187          4.9    19576      8  2011  Non Fiction

10. Вивести максимальну і  мінімальну ціну для жанру Fiction I NonFiction (за допомогою методів  groupby та agg, для підрахунку мінімальних та максимальних значень використовуйте max та min). Використовуйте тількі потрібні вам стовпці. 

In [18]:
result = df[df['genre'].isin(['Fiction', 'Non Fiction'])].groupby('genre')['price'].agg(['max', 'min'])

print("Максимальна та мінімальна ціна для жанрів Fiction та NonFiction:")
result

Максимальна та мінімальна ціна для жанрів Fiction та NonFiction:


max  min
genre                
Fiction       82    0
Non Fiction  105    0

11. Збережіть файл. Прізвище_ПР_2_2.ipynb 